{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas import read_table, concat\n",
      "from glob import glob\n",
      "\n",
      "\n",
      "tables = []\n",
      "for path in glob('res/*.pairwise_log16S_corr.tsv'):\n",
      "    tables.append(read_table(path, names=('trait', 'pairs',\n",
      "                                          'pearson_r', 'pearson_p',\n",
      "                                          'spearman_r', 'spearman_p',\n",
      "                                          'kendall_t', 'kendall_p')))\n",
      "corrs = concat(tables).set_index('trait')\n",
      "\n",
      "trait_type = []\n",
      "for trait in corrs.index:\n",
      "    if trait[0] == 'K':\n",
      "        trait_type.append('gene')\n",
      "    elif trait[0] == 'M':\n",
      "        trait_type.append('module')\n",
      "    else:\n",
      "        trait_type.append('other')\n",
      "corrs['trait_type'] = trait_type"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from matplotlib import pyplot as plt\n",
      "from matplotlib import colors, ticker\n",
      "import numpy as np\n",
      "\n",
      "fig, (ax1, ax2, ax3) = plt.subplots(3, 1, figsize=(8,12), sharex=True)\n",
      "fig.suptitle('Correlations with 16S Copy Number', y=0.94, fontsize=18)\n",
      "\n",
      "ax1.scatter(corrs.pairs, corrs.kendall_t,\n",
      "            c=np.abs(corrs.kendall_t),\n",
      "            norm=colors.PowerNorm(0.5),\n",
      "            alpha=0.05)\n",
      "ax1.set_xlim(-10, 375)\n",
      "ax1.set_ylabel(r\"Kendall's $\\tau$\")\n",
      "\n",
      "cm = ax2.scatter(corrs.pairs, corrs.kendall_p,\n",
      "                  c=np.abs(corrs.kendall_t),\n",
      "                  norm=colors.PowerNorm(0.5),\n",
      "                  alpha=0.2)\n",
      "ax2.set_yscale('symlog', linthreshy=1e-8, linscaley=1e-2)\n",
      "ax2.set_ylim(1, 0)\n",
      "ax2.set_ylabel(\"Kendall's Test p-value\")\n",
      "\n",
      "bins = np.arange(0, 400, 50)\n",
      "ax3.hist(corrs.pairs[corrs.trait_type == 'module'],\n",
      "         bins=bins, normed=True,\n",
      "         alpha=0.4, label='Modules')\n",
      "ax3.hist(corrs.pairs[corrs.trait_type == 'gene'],\n",
      "         bins=bins, normed=True,\n",
      "         alpha=0.5, label='Genes')\n",
      "ax3.set_xlabel(\"Number of Pairs\")\n",
      "ax3.legend(loc='upper right')\n",
      "\n",
      "ax1.axvline(50, color='black', linestyle='--', lw=2)   \n",
      "ax2.axvline(50, color='black', linestyle='--', lw=2)\n",
      "ax2.axhline(0.05, color='black', linestyle='--', lw=2)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# What's the relationship between the absolute number of species with a gene,\n",
      "# and the number of pairs that were found?\n",
      "\n",
      "tables = []\n",
      "for path in ['meta/K0.tsv', 'meta/M0.tsv']:\n",
      "    table = read_table(path)\n",
      "    table = table[table.columns[2:]]  # Ignore 'treelabel' and 'copies16S'\n",
      "    tables.append(table)\n",
      "traits = concat(tables, axis=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "freq = []\n",
      "nspecies = float(len(traits))\n",
      "for trt in corrs.index:\n",
      "    freq.append(sum(traits[trt] > 0) / nspecies)\n",
      "corrs['frequency'] = freq"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from matplotlib import gridspec\n",
      "\n",
      "fig = plt.figure(figsize=(6, 7))\n",
      "ax1 = plt.subplot2grid((2, 8), (0, 0), colspan=7)\n",
      "ax2 = plt.subplot2grid((2, 8), (1, 0), colspan=7, sharex=ax1)\n",
      "ax3 = plt.subplot2grid((2, 8), (0, 7), rowspan=2)\n",
      "\n",
      "#plt.subplots_adjust()\n",
      "\n",
      "\n",
      "ax1.scatter(corrs.frequency, corrs.pairs,\n",
      "            c=np.abs(corrs.kendall_p),\n",
      "            norm=colors.PowerNorm(0.4),\n",
      "            alpha=0.1)\n",
      "ax1.set_ylim(10, 400)\n",
      "ax1.set_ylabel('Number of Pairs')\n",
      "\n",
      "cm = ax2.scatter(corrs.frequency, corrs.kendall_t,\n",
      "            c=np.abs(corrs.kendall_p),\n",
      "            norm=colors.PowerNorm(0.4),\n",
      "            alpha=0.1)\n",
      "cbar = fig.colorbar(cm, cax=ax3,\n",
      "             ticks=[0.001, 0.05, 0.1],\n",
      "             label=\"Kendall's Test p-value\")\n",
      "cbar.ax.set_yticklabels(['0.001', '0.05', '0.1'])\n",
      "cbar.set_alpha(1)\n",
      "cbar.draw_all()\n",
      "ax2.set_xlim(-0.01, 1.01)\n",
      "ax2.set_ylim(-1, 1)\n",
      "ax2.set_ylabel(r\"Kendall's $\\tau$\")\n",
      "ax2.set_xlabel('Proportion of Genomes with Trait')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Of the traits that are statistically significantly correlated, what's\n",
      "# the distribution of tau?\n",
      "from matplotlib import colorbar, cm\n",
      "\n",
      "grades = 30\n",
      "cmap = plt.get_cmap(cm.gist_heat_r)\n",
      "thr = np.logspace(0, -5, grades)\n",
      "norm = colors.LogNorm()\n",
      "norm.autoscale(thr)\n",
      "\n",
      "bins = np.linspace(-1, 1, 50)\n",
      "\n",
      "fig = plt.figure()\n",
      "ax = fig.add_subplot(1, 1, 1, axisbg='gray')\n",
      "ax.set_title(r\"Fireplot Histogram\")\n",
      "for thresh in thr:\n",
      "    ax.hist(corrs.kendall_t[(corrs.pairs > 50) & (corrs.kendall_p < thresh)], bins=bins,\n",
      "            color=cmap(norm(thresh)),\n",
      "            histtype='stepfilled', lw=0)\n",
      "\n",
      "ax.set_xlabel(r\"Kendall's $\\tau$\")\n",
      "m = cm.ScalarMappable(cmap=cmap, norm=norm)\n",
      "m.set_array(thr)\n",
      "\n",
      "fig.colorbar(m, ax=ax, label=\"Kendall's Test p-value Threshold\")\n",
      "#cb = colorbar.ColorbarBase(ax, cmap=cmap, norm=norm, orientation='vertical')\n",
      "    \n",
      "# I call this a fire plot.\n",
      "# It's overlayed histograms, or density plots, colored by some threshold property.\n",
      "# You can choose an iso-hue to trace a sub-histogram of that "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fire_plot(x, y, bins=10):\n",
      "    \"\"\"Plot a fireplot histogram.\n",
      "    \n",
      "    Compute and draw a fireplot of *x* graded by y.\n",
      "    \n",
      "    \"\"\"\n",
      "    pass"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%load scripts/pairwise_log16S_corr.py"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python3\n",
      "\n",
      "from sys import argv, stdout\n",
      "from Bio.Phylo import read as read_tree\n",
      "from pandas import read_table, DataFrame\n",
      "from itertools import permutations\n",
      "from scipy.stats import pearsonr, spearmanr, kendalltau\n",
      "import numpy as np\n",
      "from numpy import log2\n",
      "import random\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def _pair_that_differ(mapping):\n",
      "    \"\"\"Select two keys in a dictionary which differ for all values.\n",
      "\n",
      "    *mapping* is a dictionary of labels mapped to a sequence of values.\n",
      "    \n",
      "    Find two keys in dictionary *mapping* for which each index in the\n",
      "    associated sequence are not equal.\n",
      "    \n",
      "    \"\"\"\n",
      "    if len(set(mapping.values())) == 1:\n",
      "        # If only one value tuple in all of the leaves, there's no mixed pair.\n",
      "        return None\n",
      "    for key_series in zip(*mapping.values()):\n",
      "        if len(set(key_series)) == 1:\n",
      "            # If any value series has only one values, then\n",
      "            # there are no mixed pairs.\n",
      "            return None\n",
      "    # If these things aren't true, then let's search for a mixed pair.\n",
      "    # The keys have to be sorted here, or else the result is non-deterministic.\n",
      "    for key1, key2 in permutations(sorted(mapping.keys()), 2):\n",
      "        seq1, seq2 = mapping[key1], mapping[key2]\n",
      "        if np.all([val1 != val2 for val1, val2 in zip(seq1, seq2)]):\n",
      "            return key1, key2\n",
      "    # TODO: Refactor so that the pairs are returned with their values,\n",
      "    # therefore values will only have to be looked up once.\n",
      "\n",
      "def _recur_mixed_pairs(clade, *key_funcs):\n",
      "    \"\"\"Worker function for *mixed_pairs*.\n",
      "\n",
      "    The function is applied recursively to a node, returning pairs and\n",
      "    independent branches from above that node.\n",
      "\n",
      "    \"\"\"\n",
      "    if clade.is_terminal():\n",
      "        value = tuple(func(clade.name) for func in key_funcs)  # TODO: Speed this up.\n",
      "        # It's a leaf.  There are no pairs above.  Return the leaf's\n",
      "        # name mapping to the leaf's value for each key as a tuple.\n",
      "        return [], {clade.name: value}\n",
      "\n",
      "    # Collect the pairs already found above this node and any free leaves\n",
      "    # above this node.\n",
      "    pairs_above = []\n",
      "    free_leaves_above = {}\n",
      "    for subclade in clade.clades:\n",
      "        pairs, free_leaves = _recur_mixed_pairs(subclade, *key_funcs)\n",
      "        pairs_above.extend(pairs)\n",
      "        free_leaves_above.update(free_leaves)\n",
      "\n",
      "    if not free_leaves_above:\n",
      "        # The node above must have formed a pair.\n",
      "        # There's no way you're going to get another pair at this node.\n",
      "        # Just return all of the pairs formed above this point and\n",
      "        # the empty dictionary.\n",
      "        return pairs_above, free_leaves_above\n",
      "    else:\n",
      "        found_pair = _pair_that_differ(free_leaves_above)\n",
      "        if found_pair:\n",
      "            # You found a new pair at this node.  Return it\n",
      "            # at the end of the list of pairs above.  There are now no\n",
      "            # free leaves above (since they wouldn't be evolutionarily\n",
      "            # independent from the leaves in this pair), so return an\n",
      "            # empty dictionary.\n",
      "            return pairs_above + [found_pair], {}\n",
      "        else:\n",
      "            # Well, there's no mixed pair of leaves above.  Return the\n",
      "            # pairs list and all of the free leaves above this node.\n",
      "            return pairs_above, free_leaves_above\n",
      "\n",
      "def mixed_pairs(clade, *key_funcs):\n",
      "    \"\"\"Search for phylogenetically independent mixed clades.\n",
      "\n",
      "    Very shallow wrapper around *_recur_mixed_pairs*.\n",
      "\n",
      "    Find the maximum number of evolutionarily independent taxon pairs.\n",
      "    See Maddison 2000, J. theor. Bio.\n",
      "\n",
      "    If multiple key functions are given, all of them must be unequal for nodes\n",
      "    to be considered mixed.\n",
      "    If you would instead like any difference to indicate a mixed pair,\n",
      "    feed a single key which returns a tuple of values.\n",
      "\n",
      "    `mixed_pairs(clade, lambda x: data[trait1][x],\n",
      "                        lambda x: data[trait2][x])`\n",
      "    will give a different result than\n",
      "    `mixed_pairs(clade, lambda x: (data[trait1][x],\n",
      "                                   data[trait2][x]))`\n",
      "    since the second key returns a tuple of values which are compared\n",
      "    as a whole, rather than being unpacked by the comparison routine.\n",
      "\n",
      "    \"\"\"\n",
      "    return _recur_mixed_pairs(clade, *key_funcs)[0]\n",
      "\n",
      "\n",
      "def contrast_pairs(trait_table, pairs):\n",
      "    tax1s, tax2s = zip(*pairs)\n",
      "    data = {}\n",
      "    for trait in trait_table.columns:\n",
      "        tax1_values = trait_table[trait].ix[list(tax1s)].values\n",
      "        tax2_values = trait_table[trait].ix[list(tax2s)].values\n",
      "        data[trait] = tax1_values - tax2_values\n",
      "    return DataFrame(data, index=pairs)\n",
      "\n",
      "def trait_corr(trait_table, clade):\n",
      "    \"\"\"Calculate the correlation between two traits based on independent pairs.\n",
      "    \n",
      "    Takes a DataFrame of two traits, indexed by the names of leaves on\n",
      "    the tree and a Biopython tree object.\n",
      "    \n",
      "    \"\"\"\n",
      "    methods = [pearsonr, spearmanr, kendalltau]\n",
      "\n",
      "    trait1, trait2 = trait_table.columns\n",
      "    make_key_func = lambda trait: (lambda x: trait_table[trait][x])\n",
      "    key_func1 = make_key_func(trait1)\n",
      "    key_func2 = make_key_func(trait2)\n",
      "\n",
      "    pairs = mixed_pairs(clade, key_func1, key_func2)\n",
      "    output = {'pairs': len(pairs)}\n",
      "    if len(pairs) == 0:\n",
      "        return output\n",
      "\n",
      "    contrast = contrast_pairs(trait_table, pairs)\n",
      "\n",
      "    if np.any(contrast.abs().sum() == 0):\n",
      "        for meth in methods:\n",
      "            output[meth.__name__] = float('nan')\n",
      "    else:\n",
      "        for meth in methods:\n",
      "            output[meth.__name__] = meth(contrast[trait1],\n",
      "                                         contrast[trait2])\n",
      "    return output\n",
      "\n",
      "def permute_data(table, cols):\n",
      "    \"\"\"Permute columns of a DataFrame.\"\"\"\n",
      "    columns = table.columns[cols]\n",
      "    table = table.copy()\n",
      "    table[columns] = table.apply(np.random.permutation)[columns]\n",
      "    return table\n",
      "\n",
      "def null_corr(trait_table, clade, cols=[2]):\n",
      "    \"\"\"Pull from the null distribution of correlations.\n",
      "    \n",
      "    Performs a permutation on traits to get at the null distribution of\n",
      "    correlation scores.\n",
      "\n",
      "    *cols* is the index of the column to shuffle:\n",
      "    0: the labels\n",
      "    1: the first column\n",
      "    2: the second column\n",
      "    \n",
      "    \"\"\"\n",
      "    original_columns = trait_table.columns\n",
      "    trait_table = trait_table.reset_index()\n",
      "    trait_table = permute_data(trait_table, cols)\n",
      "    trait_table.index = trait_table.treelabel\n",
      "    trait_table = trait_table[original_columns]\n",
      "    return trait_corr(trait_table, clade)\n",
      "\n",
      "def main():\n",
      "    tree = read_tree(argv[1], 'newick', rooted=True)\n",
      "    characters = read_table(argv[2])\n",
      "    characters.index = characters.treelabel\n",
      "    characters['log_copies16S'] = np.log2(characters.copies16S)\n",
      "    trt1 = 'log_copies16S'\n",
      "    traits = characters.columns[2:-1]\n",
      "\n",
      "    data = []\n",
      "    for trt2 in traits:\n",
      "        result = trait_corr(characters[[trt1, trt2]], tree.clade)\n",
      "        pairs = result.get('pairs')\n",
      "        pearson_r, pearson_p   = result.get('pearsonr',\n",
      "                                            tuple([float('nan')]*2))\n",
      "        spearman_r, spearman_p = result.get('spearmanr',\n",
      "                                            tuple([float('nan')]*2))\n",
      "        kendall_t, kendall_p   = result.get('kendalltau',\n",
      "                                            tuple([float('nan')]*2))\n",
      "        output = \"\\t\".join([str(x) for x in\n",
      "                           [trt2       , pairs,\n",
      "                            pearson_r  , pearson_p,\n",
      "                            spearman_r , spearman_p,\n",
      "                            kendall_t  , kendall_p]])\n",
      "        stdout.write(output)\n",
      "        stdout.write(\"\\n\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_corrs = corrs[corrs.trait_type == 'gene']\n",
      "tree = read_tree('tre/bacteria.names.nwk', 'newick', rooted=True)\n",
      "gene_table = read_table('meta/K0.tsv')\n",
      "gene_table = gene_table.set_index('treelabel')\n",
      "gene_table['log_copies16S'] = np.log2(gene_table.copies16S)\n",
      "\n",
      "genes = list(gene_corrs.index)\n",
      "trt1 = 'log_copies16S'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bins = np.linspace(0, 1, 50)\n",
      "n = 1000\n",
      "trait1, trait2 = trt1, 'K00001'\n",
      "\n",
      "# Sample of p-values based on re-picking pairs with the second trait (not 16S)\n",
      "# permuted over the phylogeny.\n",
      "# This means we've kept the phylogenetic signal for copies16S, but nullified\n",
      "# the phylogenetic AND trait correlation signal with trt2\n",
      "pick_nulls = [null_corr(gene_table[[trait1, trait2]], tree.clade)['kendalltau'][1] for _ in range(n)]\n",
      "\n",
      "\n",
      "# Now let's do a single run of making a contrast table (just one instance of pairs)\n",
      "trait_table = gene_table[[trait1, trait2]]\n",
      "make_key_func = lambda trait: (lambda x: trait_table[trait][x])\n",
      "key_func1 = make_key_func(trait1)\n",
      "key_func2 = make_key_func(trait2)\n",
      "pairs = mixed_pairs(tree.clade, key_func1, key_func2)\n",
      "trait_table = gene_table[[trait1, trait2]]\n",
      "contrast = contrast_pairs(trait_table, pairs)\n",
      "\n",
      "# And then collect a sample of p-values for the kendall test after permuting the second trait.\n",
      "# This removes the correlation between the contrasts, but keeps the magnitude of the differences.\n",
      "perm_nulls = []\n",
      "for _ in range(n):\n",
      "    kendall_t, kendall_p = kendalltau(contrast[trait1], random.sample(contrast[trait2], len(contrast)))\n",
      "    perm_nulls.append(kendall_p)\n",
      "\n",
      "fig, ax = plt.subplots(1,1)\n",
      "ax.set_title(\"Bias? ('K00001')\")\n",
      "\n",
      "ax.hist(pick_nulls, bins=bins, alpha=0.5, stacked=True, histtype='stepfilled', normed=True, label=\"Permuted Traits\")\n",
      "ax.hist(perm_nulls, bins=bins, alpha=0.5, stacked=True, histtype='stepfilled', normed=True, label=\"Permuted Contrasts\")\n",
      "ax.set_xlabel(\"p-value\")\n",
      "ax.set_ylabel(\"Frequency\")\n",
      "ax.yaxis.set_ticklabels([])\n",
      "ax.legend()\n",
      "# So, none of the bias can be explained by the values input to the correlation, and\n",
      "# all of it can be attributed to the picking of pairs."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sum(np.array(perm_nulls) < 0.001) / float(len(perm_nulls))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Since the bias is due to picking pairs, what can I do about it?\n",
      "I could try picking pairs a different way (with changes in _either_ trait denoting a \"mixed\" pair),\n",
      "but that might remove much of my signal for correlations between the two."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_hits = corrs[(corrs.trait_type == 'gene') & (corrs.kendall_p < 0.001) & (corrs.pairs >= 50)]\n",
      "null_ps = [null_corr(gene_table[[trt1, trt2]], tree.clade)['kendalltau'][1]\n",
      "           for trt2 in list(gene_hits.index) * 2]\n",
      "\n",
      "fig, ax = plt.subplots(1,1)\n",
      "ax.set_title(\"Bias in Highly Significant Hits\")\n",
      "ax.hist(gene_hits.kendall_p, bins=np.linspace(0, 1, 50), histtype='stepfilled',\n",
      "        alpha=0.25, normed=True, label='Highly Significant Hits')\n",
      "ax.hist(null_ps, bins=np.linspace(0, 1, 50), histtype='stepfilled',\n",
      "        alpha=1, normed=True, label='Permuted Traits')\n",
      "ax.set_xlabel(\"p-value\")\n",
      "ax.set_ylabel(\"Frequency\")\n",
      "ax.yaxis.set_ticklabels([])\n",
      "ax.legend()\n",
      "# The point here is that SOME of the low p-values can be explained by bias in the method\n",
      "# not not much at all."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sum(1 for x in null_ps if x < 0.001) / float(len(null_ps))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from scipy.stats import poisson\n",
      "from scipy.stats import pearsonr, spearmanr, kendalltau\n",
      "# Do Spearman and Kendall differ when you transform in a way that preserves order?\n",
      "\n",
      "x = np.random.randint(2, 15, 1000)\n",
      "x_null = np.random.randint(2, 15, 1000)\n",
      "y = poisson.rvs(np.log(x) + 0.5, size=1000)\n",
      "\n",
      "plt.scatter(x, y, alpha=0.1)\n",
      "print(pearsonr(x, y))\n",
      "print(spearmanr(x, y))\n",
      "print(kendalltau(x, y))\n",
      "print\n",
      "print(pearsonr(np.log(x), y))\n",
      "print(spearmanr(np.log(x), y))\n",
      "print(kendalltau(np.log(x), y))\n",
      "\n",
      "# The answer is no.  As long as the function is monotonic and continuous\n",
      "# then all of the correlations EXCEPT FOR PEARSON, come out the same.\n",
      "\n",
      "# That's why we talk about these types of non-parametric approaches\n",
      "# as treating the data with more honesty about our ignorance."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_hits = corrs[(corrs.trait_type == 'gene') & (corrs.kendall_p < 0.001) & (corrs.pairs >= 50)]\n",
      "null_ts = [null_corr(gene_table[[trt1, trt2]], tree.clade)['kendalltau'][0]\n",
      "           for trt2 in list(gene_hits.index) * 2]\n",
      "\n",
      "_ = plt.hist([gene_hits.kendall_t, null_ts], bins=np.linspace(-1, 1, 25), histtype='stepfilled', alpha=0.5, normed=True)\n",
      "# The point here is that SOME of the low p-values can be explained by bias in the method\n",
      "# not not much at all."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "null_ts += [null_corr(gene_table[[trt1, trt2]], tree.clade)['kendalltau'][0]\n",
      "            for trt2 in list(gene_hits.index)]\n",
      "plt.hist(null_ts, bins=np.linspace(-0.25, 0.25, 50))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from scipy.stats import shapiro, anderson\n",
      "\n",
      "shapiro(null_ts), anderson(null_ts)\n",
      "# So it doesn't look normal by either test, but it's not too far off.\n",
      "# I'll bet what we're seeing is the kurtosis/multimodality introduced by being a mixture of\n",
      "# multiple other distributions.  If I did this with just one trait, I bet we'd get a different answer."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_hits.head()  # K00012 it is!"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "trt2 = 'K00012'\n",
      "null_ts += [null_corr(gene_table[[trt1, trt2]], tree.clade)['kendalltau'] for _ in range(100)]\n",
      "_ = plt.hist(zip(*null_ts)[0], bins=np.linspace(-0.25, 0.25, 50))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ts = zip(*null_ts)[0]\n",
      "shapiro(ts), anderson(ts)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# NOW IT'S NORMAL!!"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}